// @azure/keyvault-certificates/dist-esm/keyvault-certificates/src/lro/delete/operation.js
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
import { KeyVaultCertificatePollOperation, } from "../keyVaultCertificatePoller";
import { getDeletedCertificateFromDeletedCertificateBundle } from "../../transformations";
import { createTraceFunction } from "../../../../keyvault-common/src";
/**
 * @internal
 */
const withTrace = createTraceFunction("Azure.KeyVault.Certificates.DeleteCertificatePoller");
/**
 * An interface representing a delete certificate's poll operation
 */
export class DeleteCertificatePollOperation extends KeyVaultCertificatePollOperation {
    constructor(state, vaultUrl, client, operationOptions = {}) {
        super(state, { cancelMessage: "Canceling the deletion of a certificate is not supported." });
        this.state = state;
        this.vaultUrl = vaultUrl;
        this.client = client;
        this.operationOptions = operationOptions;
    }
    /**
     * The DELETE operation applies to any certificate stored in Azure Key Vault. DELETE cannot be applied
     * to an individual version of a certificate. This operation requires the certificates/delete permission.
     */
    deleteCertificate(certificateName, options = {}) {
        return withTrace("deleteCertificate", options, async (updatedOptions) => {
            const response = await this.client.deleteCertificate(this.vaultUrl, certificateName, updatedOptions);
            return getDeletedCertificateFromDeletedCertificateBundle(response);
        });
    }
    /**
     * Retrieves the deleted certificate information plus its attributes, such as retention interval, scheduled permanent deletion and the
     * current deletion recovery level. This operation requires the certificates/get permission.
     */
    async getDeletedCertificate(certificateName, options = {}) {
        return withTrace("getDeletedCertificate", options, async (updatedOptions) => {
            const result = await this.client.getDeletedCertificate(this.vaultUrl, certificateName, updatedOptions);
            return getDeletedCertificateFromDeletedCertificateBundle(result._response.parsedBody);
        });
    }
    /**
     * Reaches to the service and updates the delete certificate's poll operation.
     */
    async update(options = {}) {
        const state = this.state;
        const { certificateName } = state;
        if (options.abortSignal) {
            this.operationOptions.abortSignal = options.abortSignal;
        }
        if (!state.isStarted) {
            const deletedCertificate = await this.deleteCertificate(certificateName, this.operationOptions);
            state.isStarted = true;
            state.result = deletedCertificate;
            if (!deletedCertificate.recoveryId) {
                state.isCompleted = true;
            }
        }
        if (!state.isCompleted) {
            try {
                state.result = await this.getDeletedCertificate(certificateName, this.operationOptions);
                state.isCompleted = true;
            }
            catch (error) {
                if (error.statusCode === 403) {
                    // At this point, the resource exists but the user doesn't have access to it.
                    state.isCompleted = true;
                }
                else if (error.statusCode !== 404) {
                    state.error = error;
                    state.isCompleted = true;
                    throw error;
                }
            }
        }
        return this;
    }
}
